File: Easy/two_number_sum.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <unordered_map>
#include <bits/stdc++.h>
using namespace std;

vector<int> twoNumberSum(vector<int> array, int targetSum)
{
    // Write your code here.
    vector<int> result;
    const int number_of_elements = array.size();

    /* 	this solution takes O(n^2) time
	for (int i = 0; i < number_of_elements - 1; i++)
	{
		for (int j = i + 1; j < number_of_elements; j++)
		{
			if ((array[i] + array[j]) == targetSum)
			{
				result.push_back(array[i]);
				result.push_back(array[j]);
			}
		}
	}
*/

    // unordered_map<int, int> counts;
    // for (int i = 0; i < number_of_elements; i++)
    // {
    //     counts[array[i]]++;
    // }
    /*
    cout << "displaying map\n";
    unordered_map<int, int>::iterator itr = counts.begin();
    while (itr != counts.end())
    {
        cout << itr->first << " " << itr->second << '\n';
        itr++;
    }
*/
    /*
    for (int i = 0; i < number_of_elements; i++)
    {
        int current_value = array[i];
        int remainder_needed = targetSum - current_value;
        cout << endl;
        if (current_value != remainder_needed)
        {
            if (counts[remainder_needed] > 0)
            {
                return vector<int>{current_value, remainder_needed};
            }
        }

        if (current_value == remainder_needed)
        {
            if (counts[current_value] > 1)
            {
                return vector<int>{current_value, current_value};
            }
        }
    }
    */

    return {};
}

vector<int> twoNumberSumTwo(vector<int> &array, int targetSum)
{
    int n = array.size();
    // O(n log n)
    sort(array.begin(), array.end());
    int left = 0;
    int right = n - 1;
    // O ( log n)
    while (left < right)
    {
        int current_sum = array[left] + array[right];
        if (current_sum == targetSum)
            return vector<int>{array[left], array[right]};
        else if (current_sum > targetSum)
            right--;
        else
            left++;
    }

    return {};
}

void display(const vector<int> &result)
{
    cout << '\n';
    for (int i = 0; i < result.size(); i++)
        cout << result[i] << " ";
}

int main()
{
    int number_of_elements, targetSum;
    cin >> number_of_elements >> targetSum;

    vector<int> array(number_of_elements);
    for (int i = 0; i < number_of_elements; i++)
        cin >> array[i];

    vector<int> result = twoNumberSumTwo(array, targetSum);
    display(result);

    return EXIT_SUCCESS;
}

File: Easy/two_number_sum.py
number_of_elements = input()
target_sum = input()
array = []

number_of_elements = int(number_of_elements)
target_sum = int(target_sum)

for i in range(number_of_elements):
    temp = int(input())
    array.append(temp)


def twoNumberSum(array, target_sum):
    nums = {}
    for num in array:
        potentialMatch = target_sum - num
        if potentialMatch in nums:
            return [num, potentialMatch]
        else:
            nums[num] = True

    return []


# result = twoNumberSum(array, target_sum)
# for num in result:
#     print(num, end=" ")

"""
def bruteForceSum(array, target_sum):
    for i in range(len(array) - 1):
        for j in range(i + 1, len(array)):
            if array[i] + array[j] == target_sum:
                return [array[i], array[j]]

    return []


result = bruteForceSum(array, target_sum)
for num in result:
    print(num, end=" ")
"""


def sortingSearch(array, target_sum):
    # array = sorted(array)
    array.sort()
    left = 0
    right = len(array) - 1
    while left < right:
        current_sum = array[left] + array[right]
        if current_sum == target_sum:
            return [array[left], array[right]]
        elif current_sum > target_sum:
            right = right - 1
        else:
            left = left + 1

    return []


result = sortingSearch(array, target_sum)
for num in result:
    print(num, end=" ")
